crates/transport/src/layers/fallback.rs

```diff
@@ -0,0 +1,453 @@
+    ///
+    /// - The `active_transport_count` parameter controls how many transports are used for requests
+    ///   at any one time.
+    /// - The `log_transport_rankings` parameter controls whether the current transport rankings are
+    ///   output as traces. In order to see them, make sure to enable the `alloy_transport` target
+    ///   (i.e. `RUST_LOG=alloy_transport=debug`).
+    pub fn new(
+        transports: Vec<S>,
+        active_transport_count: usize,
+        log_transport_rankings: bool,
+    ) -> Self {
+        let scored_transports = transports
+            .into_iter()
+            .enumerate()
+            .map(|(id, transport)| ScoredTransport::new(id, transport))
+            .collect::<Vec<_>>();
+
+        Self {
+            transports: Arc::new(RwLock::new(scored_transports)),
+            active_transport_count,
+            log_transport_rankings,
+        }
+    }
+
+    /// Log the current ranking of transports
+    fn log_transport_rankings(&self)
+    where
+        S: Debug,
+    {
+        let transports = self.transports.read();
+
+        debug!("Current transport rankings:");
+        for (idx, transport) in transports.iter().enumerate() {
+            debug!("  #{}: Transport[{}] - {}", idx + 1, transport.id, transport.metrics_summary());
+        }
+    }
+}
+
+impl<S> FallbackService<S>
+where
+    S: Service<RequestPacket, Future = TransportFut<'static>, Error = TransportError>
+        + Send
+        + Clone
+        + Debug
+        + 'static,
+{
+    /// Make a request to the fallback service middleware.
+    ///
+    /// Here is a high-level overview of how requests are handled:
+    ///
+    /// - At the start of each request, we assume `self.transports` to be sorted by score
+    /// - We take the top `self.active_transport_count` transports and call them in parallel
+    /// - If any of them succeeds, we re-sort `self.transports` to reflect the new scores
+    /// - We then return the first successful response and clear the remaining futures
+    /// - If all transports fail, we return the last error that occurred
+    ///
+    /// This strategy allows us to always make requests to the best available transports
+    /// while keeping them available. We also only update the transport scores after receiving
+    /// a response, avoiding unnecessary re-sorting.
+    async fn make_request(&self, req: RequestPacket) -> Result<ResponsePacket, TransportError> {
+        // Get the top transports to use, without removing them from the heap
+        let top_transports = {
+            let transports = self.transports.read();
+            transports.iter().take(self.active_transport_count).cloned().collect::<Vec<_>>()
+        };
+
+        // Create a collection of future requests
+        let mut futures = FuturesUnordered::new();
+
+        // Launch requests to all active transports in parallel
+        for transport in top_transports {
+            let start = Instant::now();
```

Comment:
this instant init should be moved into the async move, otherwise this doesn't accurately track the time this future takes